### PR TITLE
Stac index issue with sentinel 2

### LINKED ISSUE #510: Cannot index alternate Sentinel-2 product
I am attempting to index Sentinel-2 data (DEAustralia's `ga_s2am_ard_3`) using a STAC document. However it seems like it's having trouble handling it instead of [DEAfrica's `s2_l2a`](https://explorer.digitalearth.africa/products/s2_l2a). 

```
# s3-to-dc --stac s3://dea-public-data-dev/baseline/ga_s2am_ard_3/53/KQB/2022/01/10/20220110T022746/ga_s2am_ard_3-2-1_53KQB_2022-01-10_final.stac-item.json ga_s2am_ard_3
Added 0 datasets, skipped 0 datasets and failed 0 datasets.
```

Interesting how `s3-to-dc` didn't pick anything up, however `sqs-to-dc` did give a more detailed error.

```
# aws s3 cp s3://dea-public-data-dev/baseline/ga_s2am_ard_3/53/KQB/2022/01/10/20220110T022746/ga_s2am_ard_3-2-1_53KQB_2022-01-10_final.stac-item.json .

# aws sns publish --region ap-southeast-2 --topic-arn arn:aws:sns:ap-southeast-2:xxxxxxxxxxxx:stac-testing-topic --message file://ga_s2am_ard_3-2-1_53KQB_2022-01-10_final.stac-item.json

# sqs-to-dc --skip-lineage --stac stac-testing-queue ga_s2am_ard_3
ERROR:root:Failed to handle message with exception: Failed to create dataset with error Dataset metadata did not match product signature.
Dataset definition:
 {
    "product": {
        "name": "s2_l2a"
    },
    "properties": {
        "title": "ga_s2am_ard_3-2-1_53KQB_2022-01-10_final",
        "dea:dataset_maturity": "final",
        "dea:product_maturity": "stable",
        "eo:cloud_cover": 60.72365718760058,
        "eo:constellation": "sentinel-2",
        "eo:gsd": 10.0,
        "eo:instrument": "MSI",
        "eo:platform": "sentinel-2a",
        "eo:sun_azimuth": 106.059403157537,
        "eo:sun_elevation": 25.4820222426119,
        "fmask:clear": 38.286302965152736,
        "fmask:cloud": 60.72365718760058,
        "fmask:cloud_shadow": 0.8214239501527865,
        "fmask:snow": 0.0005308542440137889,
        "fmask:water": 0.168085042849891,
        "gqa:abs_iterative_mean_x": 0.3,
        "gqa:abs_iterative_mean_xy": 0.4,
        "gqa:abs_iterative_mean_y": 0.26,
        "gqa:abs_x": 0.4,
        "gqa:abs_xy": 0.55,
        "gqa:abs_y": 0.38,
        "gqa:cep90": 0.69,
        "gqa:iterative_mean_x": 0.27,
        "gqa:iterative_mean_xy": 0.31,
        "gqa:iterative_mean_y": 0.15,
        "gqa:iterative_stddev_x": 0.21,
        "gqa:iterative_stddev_xy": 0.37,
        "gqa:iterative_stddev_y": 0.3,
        "gqa:mean_x": 0.26,
        "gqa:mean_xy": 0.31,
        "gqa:mean_y": 0.16,
        "gqa:stddev_x": 0.93,
        "gqa:stddev_xy": 1.32,
        "gqa:stddev_y": 0.94,
        "odc:dataset_version": "3.2.1",
        "odc:file_format": "GeoTIFF",
        "created": "2022-05-26T12:19:44.733764Z",
        "odc:producer": "ga.gov.au",
        "odc:product_family": "ard",
        "odc:region_code": "53KQB",
        "s2cloudless:clear": 43.367463943384394,
        "s2cloudless:cloud": 56.6325360566156,
        "sat:orbit_state": "descending",
        "sat:relative_orbit": 45,
        "sentinel:datastrip_id": "S2A_OPER_MSI_L1C_DS_VGS1_20220110T022746_S20220110T011116_N03.01",
        "sentinel:datatake_start_datetime": "2022-01-10T02:27:46Z",
        "sentinel:product_name": "S2A_MSIL1C_20220110T010731_N0301_R045_T53KQB_20220110T022746",
        "sentinel:sentinel_tile_id": "S2A_OPER_MSI_L1C_TL_VGS1_20220110T022746_A034219_T53KQB_N03.01",
        "proj:epsg": 32753,
        "proj:shape": [
            5490,
            5490
        ],
        "proj:transform": [
            20.0,
            0.0,
            699960.0,
            0.0,
            -20.0,
            8200000.0,
            0.0,
            0.0,
            1.0
        ],
        "datetime": "2022-01-10T01:11:50.577728Z",
        "odc:processing_datetime": "2022-05-26T12:19:44.733764Z"
    }
}

Product signature:
 {
    "product": {
        "name": "ga_s2am_ard_3"
    },
    "properties": {
        "eo:platform": "sentinel-2a",
        "odc:producer": "ga.gov.au",
        "eo:instrument": "MSI",
        "odc:product_family": "ard",
        "dea:product_maturity": "stable"
    }
}

 The URI was s3://dea-public-data-dev/baseline/ga_s2am_ard_3/53/KQB/2022/01/10/20220110T022746/ga_s2am_ard_3-2-1_53KQB_2022-01-10_final.stac-item.json
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/odc/apps/dc_tools/sqs_to_dc.py", line 246, in queue_to_odc
    index_update_dataset(
  File "/usr/local/lib/python3.8/dist-packages/odc/apps/dc_tools/utils.py", line 199, in index_update_dataset
    raise IndexingException(
odc.apps.dc_tools.utils.IndexingException: Failed to create dataset with error Dataset metadata did not match product signature.
Dataset definition:
 {
    "product": {
        "name": "s2_l2a"
    },
    "properties": {
        "title": "ga_s2am_ard_3-2-1_53KQB_2022-01-10_final",
        "dea:dataset_maturity": "final",
        "dea:product_maturity": "stable",
        "eo:cloud_cover": 60.72365718760058,
        "eo:constellation": "sentinel-2",
        "eo:gsd": 10.0,
        "eo:instrument": "MSI",
        "eo:platform": "sentinel-2a",
        "eo:sun_azimuth": 106.059403157537,
        "eo:sun_elevation": 25.4820222426119,
        "fmask:clear": 38.286302965152736,
        "fmask:cloud": 60.72365718760058,
        "fmask:cloud_shadow": 0.8214239501527865,
        "fmask:snow": 0.0005308542440137889,
        "fmask:water": 0.168085042849891,
        "gqa:abs_iterative_mean_x": 0.3,
        "gqa:abs_iterative_mean_xy": 0.4,
        "gqa:abs_iterative_mean_y": 0.26,
        "gqa:abs_x": 0.4,
        "gqa:abs_xy": 0.55,
        "gqa:abs_y": 0.38,
        "gqa:cep90": 0.69,
        "gqa:iterative_mean_x": 0.27,
        "gqa:iterative_mean_xy": 0.31,
        "gqa:iterative_mean_y": 0.15,
        "gqa:iterative_stddev_x": 0.21,
        "gqa:iterative_stddev_xy": 0.37,
        "gqa:iterative_stddev_y": 0.3,
        "gqa:mean_x": 0.26,
        "gqa:mean_xy": 0.31,
        "gqa:mean_y": 0.16,
        "gqa:stddev_x": 0.93,
        "gqa:stddev_xy": 1.32,
        "gqa:stddev_y": 0.94,
        "odc:dataset_version": "3.2.1",
        "odc:file_format": "GeoTIFF",
        "created": "2022-05-26T12:19:44.733764Z",
        "odc:producer": "ga.gov.au",
        "odc:product_family": "ard",
        "odc:region_code": "53KQB",
        "s2cloudless:clear": 43.367463943384394,
        "s2cloudless:cloud": 56.6325360566156,
        "sat:orbit_state": "descending",
        "sat:relative_orbit": 45,
        "sentinel:datastrip_id": "S2A_OPER_MSI_L1C_DS_VGS1_20220110T022746_S20220110T011116_N03.01",
        "sentinel:datatake_start_datetime": "2022-01-10T02:27:46Z",
        "sentinel:product_name": "S2A_MSIL1C_20220110T010731_N0301_R045_T53KQB_20220110T022746",
        "sentinel:sentinel_tile_id": "S2A_OPER_MSI_L1C_TL_VGS1_20220110T022746_A034219_T53KQB_N03.01",
        "proj:epsg": 32753,
        "proj:shape": [
            5490,
            5490
        ],
        "proj:transform": [
            20.0,
            0.0,
            699960.0,
            0.0,
            -20.0,
            8200000.0,
            0.0,
            0.0,
            1.0
        ],
        "datetime": "2022-01-10T01:11:50.577728Z",
        "odc:processing_datetime": "2022-05-26T12:19:44.733764Z"
    }
}

Product signature:
 {
    "product": {
        "name": "ga_s2am_ard_3"
    },
    "properties": {
        "eo:platform": "sentinel-2a",
        "odc:producer": "ga.gov.au",
        "eo:instrument": "MSI",
        "odc:product_family": "ard",
        "dea:product_maturity": "stable"
    }
}

 The URI was s3://dea-public-data-dev/baseline/ga_s2am_ard_3/53/KQB/2022/01/10/20220110T022746/ga_s2am_ard_3-2-1_53KQB_2022-01-10_final.stac-item.json
Added 0 Dataset(s), Failed 1 Dataset(s), Skipped 0 Dataset(s)
```

If I had to make a guess on the source of the issue, it's probably something to do with the [code handling mapping from STAC to EO3](https://github.com/opendatacube/odc-tools/blob/develop/apps/dc_tools/odc/apps/dc_tools/_stac.py#L110).

### PR DESCRIPTION
**Pull request by emmaai**
To fix https://github.com/opendatacube/odc-tools/issues/510

The change is to constrain the hardcoded product name to DEAfrica sentinel 2 rather than ALL sentinel 2. The hack is kept there for not interrupting DEAfrica pipeline even though it's not supposed to.

Sorry I forgot to separate the formatting and real change in different commits, again.

@SeffatChowdhury

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -153,6 +154,7 @@ jobs:
       shell: bash
       run: |
         datacube system check
+        datacube metadata add "https://raw.githubusercontent.com/GeoscienceAustralia/dea-config/master/product_metadata/eo3_sentinel_ard.odc-type.yaml"
```
Comment (jmettes) [inline .github/workflows/main.yml:157]: Any there any other metadata types added elsewhere? If not, why is this one different? If there are, should this belong there?

File: apps/dc_tools/odc/apps/dc_tools/_stac.py [hunk 2]
```diff
@@ -98,30 +99,27 @@ def _stac_product_lookup(
     if constellation is not None:
         constellation = constellation.lower().replace(" ", "-")
 
-    # Let's check for Sentinel-2, which is a special case
-    # The below section handles the Element 84 and Microsft PC STAC documents
```
Comment (jmettes) [inline apps/dc_tools/odc/apps/dc_tools/_stac.py:102]: Do these comments need to be preserved? Seems to be the justification for the hack. Ideally there'd be test cases for these special cases...

File: .github/workflows/main.yml (hunk 1 above)
Comment (emmaai) [reply .github/workflows/main.yml:157]: We require `metadata type` to add ls and s2 C3 products (of DEA). I don't find a script to gather the database init things. It's a quick fix to make sure everything is working.

File: apps/dc_tools/odc/apps/dc_tools/_stac.py (hunk 2 above)
Comment (emmaai) [reply apps/dc_tools/odc/apps/dc_tools/_stac.py:102]: I don't think it does what it claims. There is no justification to hardcode the product name. After the change, that segment was limited to DEAfrica sentinel 2 product. Hence it doesn't have any general meaning to cater special format of stac documents. Had it been the point, it should have been coded in `stac_transform` to handle the format rather a single product.

**Review by jmettes — APPROVED** [review]

**Review by SpacemanPaul — APPROVED** [review]
Would be nice if DEAfrica could do a bit of housekeeping here too.  (Such hacks!)

**Review by omad — APPROVED** [review]

**Pull request merged**